Improve autocompletion by looking on the type and name

bnjjj opened the pull request:
This tweet (https://twitter.com/tjholowaychuk/status/1248918374731714560) gaves me the idea to implement that in rust-analyzer.

Basically for this first example I made some examples when we are in a function call definition. I look on the parameter list to prioritize autocompletions for the same types and if it's the same type + the same name then it's displayed first in the completion list.

So here is a draft, first step to open a discussion and know what you think about the implementation. It works (cf tests) but maybe I can make a better implementation at some places. Be careful the code needs some refactoring to be better and concise.

PS: It was lot of fun writing this haha

lnicola commented:
I wonder if we could make a `MatchInfo` enum or struct and put these heuristics in its `PartialEq` implementation, since it looks like the comparer might end up unwieldy.

bnjjj commented:
Yes definitely we can find a better way to handle this sort. I also thought about a type which could implement PartialEq. An enum for each different cases (in a function call, in a struct literal expression, ...) with all useful data inside. For exemple the call_info and fields in this first case.

bnjjj commented:
I made a little change, I moved all the sort directly in the `Completions` struct. Then I added a `sort_option` field to `Completions`. I'm questioning about the possibility to have several sort options. Could it be possible or not ? To me it makes no sense but maybe I'm missing something

lnicola commented:
Just for the record, I was thinking of something like:

```rust
#[derive(PartialEq)]
enum CompletionMatchType {
    Prefix,
    Infix, // nMat for CompletionMatchType
    Substring, // CMT for CompletionMatchType
}

enum CompletionItemType {
    Fn(FnDef),
    Field(FieldDef),
    Macro(MacroDef),
    // ...
}

struct CompletionItem {
    ref: CompletionItemType,
    type: CompletionMatchType,
    is_type_match: bool,
    // ...
}

impl PartialEq for CompletionItem {
    // heuristics here
}
```

bnjjj commented:
Oh yes, I can implement `PartialEq for CompletionItem`  and move all my code inside the current sort function inside this implementation. But about the `CompletionMatchType` I'm not sure it's useful. Do you have a use case in mind ? I mean another use case than a prefix match.

lnicola commented:
> Do you have a use case in mind ? I mean another use case than a prefix match.

Well, I'm not sure if it's the best approach, but we might want to show completions for all of those cases: prefix, literal substring (sequence) and non-contiguous substrings. Most IDEs can suggest `CompletionMatchType` if you type CMT, for example. Having that enum might be an easy way to support these in the future because we could just include it in the sorting key. Of course, it might be even better to keep track of "frecency"  (frequency + recency) and include more scoring heuristics like the web browsers do, but that seems much harder to implement.

Anyway, it's fine to not add it now. It might not even fit well with the rest of the code.

bnjjj commented:
Ok I experimented a little bit. I was not very satisfied with the implementation of `Ord` on `CompletionItem` because I need to have data coming from context (`call_info` for example). And I'm not very confident to put this data multiple times in each `CompletionItem`. It's a lot of duplicate informations IMHO.

BTW as @matklad told me on Zulip, in fact even if we sort our completions in LSP server the client doesn't care about it. So I added an additional data called `sort_text` to sort completion given the current index in the vector.

bnjjj commented:
I just add the support also for `RecordField` in a struct literal expression.

matklad commented:
Has anyone found any docs about how exactly `sortText` and `filterText` interact? Am I correct that in the current PR we completely ignore the textual match of identifiers? 


Implementation wise, I feel relying on the order here would be a wrong abstraction. A *sorted* completion list tells you that there's an unambiguous linear order from worst to best, but that's not actually true in practice. We, for example, can't generally compare unrelated completions, like a reference completion vs keyword completion. 

I think the right abstraction is to add a 

```rust
struct CompltionScore {
   ...
}
```

to each completion field, where score is something with `PartialOrd`. For example, it can be an enum only same variants of which are comparable. In the current impl, that would be just

```rust
struct CompltionScore {
    is_precise_type_match: bool
}
```

but it than could be extended with all kinds of heuristics. Then, it becomes the job of `conv` to convert those scores to a representation, suitable for the client.

How to do that given the current constraints of LSP is, of course, an option question (especially, how to marry fuzz-matching scores with semantic scores). The simplest first step seems to just mark items with fully matched types as `preselect: true`?

bnjjj commented:
About your first question to be honest I just checked the GoPLS codebase https://github.com/golang/tools/blob/master/internal/lsp/completion.go#L111

Ok then for the first iteration I will make a `CompletionScore` struct like this:
```rust
struct CompletionScore {
    is_precise_type_match: bool, // Only when it's the same type
    is_precise_type_name_match: bool, // Only when it's a perfect match and in that case I put a preselect: true
}
```

And then I will try to find some documentations about what the consequence of using `sortText` with `filterText`. Seems good for you ?

matklad commented:
Sounds good

On Thu, 16 Apr 2020 at 15:32, Coenen Benjamin <notifications@github.com>
wrote:

> About your first question to be honest I just checked the GoPLS codebase
> https://github.com/golang/tools/blob/master/internal/lsp/completion.go#L111
>
> Ok then for the first iteration I will make a CompletionScore struct like
> this:
>
> struct CompletionScore {
>     is_precise_type_match: bool, // Only when it's the same type
>     is_precise_type_name_match: bool, // Only when it's a perfect match and in that case I put a preselect: true
> }
>
> And then I will try to find some documentations about what the consequence
> of using sortText with filterText. Seems good for you ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-analyzer/rust-analyzer/pull/3954#issuecomment-614655087>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANB3M6C7QJJS5EGDMAZ4Y3RM4CHXANCNFSM4MGEQ6VQ>
> .
>

bnjjj commented:
Ok I made some changes with your advices. And here is screenshot in vscode to compare between current situation in RA and new situation with this PR.

Current:
<img width="603" alt="Capture d’écran 2020-04-16 à 18 28 54" src="https://user-images.githubusercontent.com/5719034/79481905-62953880-8010-11ea-9b91-d48f5c5ab0d1.png">

New:
<img width="634" alt="Capture d’écran 2020-04-16 à 18 27 52" src="https://user-images.githubusercontent.com/5719034/79482087-98d2b800-8010-11ea-880f-d1094f4ddd41.png">

In fact I only reorder completions with a score. If not it keeps the same behavior as before (seems to be an alphabetical sort by default).

matklad commented on crates/ra_ide/src/completion/completion_item.rs:
```diff
@@ -314,6 +346,54 @@ impl Completions {
+        self.score_option = Some(score_option);
+    }
+
+    pub(crate) fn compute_score(&mut self, ctx: &CompletionContext) {
```
Why do we need to delay the computation of scores until the end of completion process? A simpler approach seems to be to add a fully-computed score to `CompletionItem`.

bnjjj replied:
Yes I will move this part in completion_dot

matklad commented on crates/rust-analyzer/src/conv.rs:
```diff
@@ -163,6 +163,14 @@ impl ConvWith<(&LineIndex, LineEndings)> for CompletionItem {
+                CompletionScore::TypeAndNameMatch => res.preselect = Some(true),
+                CompletionScore::TypeMatch => {}
+            }
+            res.sort_text = Some(format!("{:02}", ctx.2));
```
I'd rather start with just setting preselect and void the sort_text hack. I belive setitng sort_text only for some fields would be wrong, and we also need to take fuzzy mathcing into account, if we set sort text. I think just preselect would give a meaningful improvement.

bnjjj replied:
If you have an example in mind in which case sort text could cause an issue let me know I will try. The fact is as a rust analyzer user I feel better with my example in sort text. But of course I don’t want to introduce a bug or non deterministic behavior. So if you have examples on which I can make some tests I will be happy to try and show you the result. I agree with you, preselect is good but if we can’t go a step further using sortText as gopls does. Developer experience will be improved.

bnjjj replied:
By the way when I go to the LSP specs from Microsoft, when I look about documentation about `sortText` here it is `A string that should be used when comparing this item with other items. When falsy the label is used.` it seems it's not a hack. I mean, today we have all completion items sorted in alphabetical, which in my opinion, is very bad. IMHO I think we can't do something more bad because you always have to scroll the entire autocompletions list to see all interesting parameters of a struct for example. I think it really worth to at least show to the user fields with right types first + preselect perfect match. If we look at my screenshots there are keywords like `box`, `dbg`, ... displayed and at this place we don't care about it. In my example it's not so annoying because I have a lot of fields beginning with `a` letter so it's in first position and I don't have to scroll so much. If you are worried I can write a lot of tests, make a lot of experiments even manually and show you the results because I really think it worth.

matklad commented on xtask/tests/tidy-tests/main.rs:
```diff
@@ -46,9 +46,9 @@ fn check_trailing_ws(path: &Path, text: &str) {
+    for (line_number, line) in text.lines().enumerate() {
         if line.chars().last().map(char::is_whitespace) == Some(true) {
-            panic!("Trailing whitespace in {}", path.display())
+            panic!("Trailing whitespace in {} at line {}", path.display(), line_number)
```
:+1: , worth a separate PR

kjeremy commented:
I believe that `sortText` is used to literally sort in the UI so if you had:
|label|sort|
|-----|-----|
|foo|'0000'|
|bar|'0001'|

If would show up as `foo` then `bar`. If you had:

|label|sort|
|-----|-----|
|foo|'0001'|
|bar|'0000'|

It would be presented as `bar` then `foo`.

`filterText` is used when the characters types might not match the label: ex. typing a '.' to auto fill something.

If either of those are missing `label` is used instead.

I think there are vscode settings though that can alter the behavior and I don't know how they interact with the LSP.

bnjjj commented:
Yes sortText is used to sort completionItem and with my experiments it seems that it’s sorted by ascending so 0000 first and then 0001.

matklad commented on crates/ra_ide/src/completion/complete_dot.rs:
```diff
@@ -2,12 +2,16 @@
     CompletionItem,
 };
 use rustc_hash::FxHashSet;
+// use std::cmp::Ordering;
```
stray comment?

bnjjj replied:
deleted

matklad commented on crates/ra_ide/src/completion/completion_item.rs:
```diff
@@ -285,6 +304,51 @@ impl Builder {
         self.deprecated = Some(deprecated);
         self
     }
+    #[allow(unused)]
```
why is this allow unused?

bnjjj replied:
Because at this time I don't really use it. I just wanted to keep the same way than it was done with other fields. To stay consistent

matklad commented on crates/ra_ide/src/completion/completion_item.rs:
```diff
@@ -285,6 +304,51 @@ impl Builder {
         self
     }
+    #[allow(unused)]
+    pub(crate) fn compute_score(mut self, ctx: &CompletionContext) -> Builder {
```
I think this function should live in `presentation.rs`, and here we only need `set_score`

bnjjj replied:
updated

matklad commented on crates/rust-analyzer/src/main_loop/handlers.rs:
```diff
@@ -423,8 +423,11 @@ pub fn handle_completion(
-        items.into_iter().map(|item| item.conv_with((&line_index, line_endings))).collect();
+    let items: Vec<CompletionItem> = items
+        .into_iter()
+        .enumerate()
```
Do we actually sort variants anywhere? ctrl+f sort gives no results

bnjjj replied:
I only use the index to order when there is a type match. I don't know if it answers your question

matklad replied:
Hm, I am not sure....

Like, we never sort completion items, so the indexes they get are arbitrary. So, when we set an index as a `sort_text`, we get arbitrary ordering?

bnjjj replied:
yes you're right. We don't care about the index here. In fact, what's is important is to have a kind of global counter. I also don't care about the order at this time. I just use this data as a counter. If it's weird to you I also can pass an &mut counter coming from this `handle_completion` function but I need to have a kind of global counter to not put the same number in `sortText`. And as for non scored item there isn't any sortText it's the default behavior as before then it doesn't change. Then in fact I could have a `sortText` set to 2 and the next one set to 4 but it's not so important. The only important thing is to have a incremental counter. I will change this to pass a &mut counter it should be easier to read I think :)

bnjjj replied:
I updated the variable to use counter instead of arbitrary index

bnjjj commented:
Is it good for you @matklad ?

matklad commented:
bors r+

I want to merge this as it has been seeting in the queue for far too long. That said, I want to tweak this a quite a bit after it lands -- I have a strong suspicion that current usages of `sortText`, where we set it only for *some* variants, and don't look at the textual match at all, gives us essentially random completions. This suspicion is confirmde by the fact that a bona fide random sorting was only caught during code review. But the general idea here is definitelly solid
